GLMakie/src/glshaders/surface.jl

```diff
@@ -48,7 +49,8 @@ function _position_calc(
     """
     int index1D = index + offseti.x + offseti.y * dims.x + (index/(dims.x-1));
     ivec2 index2D = ind2sub(dims, index1D);
-    vec2 index01 = vec2(index2D) / (vec2(dims)-1.0);
```

Comment:
Hm, it's this change that makes the test fail.. It looks a bit like the colors get moved out of the surface:
PR:
<img width="399" alt="image" src="https://user-images.githubusercontent.com/1010467/228648223-3c98840a-6652-4948-9d0f-01d8282335aa.png">
With the old `vec2(index2D) / (vec2(dims)-1.0)`:
<img width="399" alt="image" src="https://user-images.githubusercontent.com/1010467/228648344-826aa527-85b8-4f74-a279-f41651afc20a.png">

Considering, that the old behavior maps more directly to the displayed matrix, why was this changed?

Reply:
With `vec2(index2D) / (vec2(dims)-1.0)` you get values between 0 and 1, so the first index will sample the left edge of a pixel in a texture and the last index will sample the right edge of a pixel.

With `vec2(index2D) / (vec2(dims))` you'd always sample the left edge of a pixel. With the `0.5` shift that becomes the center of a pixel. 

I think I made this change to fix the discrepancy between GLMakies generated mesh and the wireframe representation of CairoMakies mesh: (right)
![img](https://user-images.githubusercontent.com/10947937/212936134-f2faa5c4-d251-450f-9958-876d3834eb5a.png)

Reply:
I think is also an expected/desired change.

The color data we have here is:
```
 1.0  NaN        NaN        5.0
 2.5    2.66667    2.83333  3.0
```
In surface these should map directly to vertices and we should be interpolating between them. Assuming quad faces, this should generate all NaN's. 

With vertices marked as `+` and the positions where the texture is sampled without interpolation marked with pink dots, this should be the before:
![Screenshot from 2023-03-30 00-29-11](https://user-images.githubusercontent.com/10947937/228682674-182ebd97-b979-4c37-8a5b-8127246ae5ce.png)
And this the after:
![Screenshot from 2023-03-30 00-29-21](https://user-images.githubusercontent.com/10947937/228682689-1f80ec7d-565c-4976-a67d-c9059bad1290.png)

Reply:
Hm, but we shouldn't render the edges pretty much outside the visible field?
I think, if we do this, we'll need to move the interpolation center to the middle of the quad!

Reply:
Maybe to get this branch merged, revert things for now, so that we dont have a regression?

Reply:
My understanding of surface plots is that it's a mesh generating method where vertex positions follow from xs and ys and the colors/values passed should exactly map to those positions. That's not the case on master. That's a bug imo.

Moving the color to the center of a quad doesn't work because we have N colors for N vertices. We could move the full vertex to quad center but then we would need to generate an N+1 grid to keep quad sizes regular.

Maybe it would be useful to compare to some other plotting libraries? To me it should just be like my after example.

Reply:
If we look at this with automatically chosen color values, i.e. z values, I would expect the first color in the colormap to directly correspond to the lowest z value, the last to the highest and anything in between  to be smoothly interpolated. That's not the case on master, but is the case in the pr. 

This example makes the difference very obvious:
```julia
scene = Scene()
surface!(scene, [1 2 3; 1 2 3; 1 2 3], colormap = [:black, :white, :black], shading = false)
cam3d!(scene, projectiontype = Makie.Orthographic)
center!(scene)
axis3d!(scene)
lines!(scene, Point3f[(3, 3, 1), (3, 3, 2), (3, 3, 3)], color = [:black, :white, :black], linewidth = 50)
update_cam!(scene, Vec3f(5.5, 5, 2), Vec3f(1.5, 1.5, 2))
scene
```

![Screenshot 2023-04-05 154927](https://user-images.githubusercontent.com/10947937/230100860-6d23ef80-1aed-4cc9-a609-915a0d711598.png)

The line to the left reaches full black at the line start/end, i.e. at minimum and maximum z. The surface on the other hand at z values significantly larger than the minimum and significantly lower than the maximum.

On another note, why is `surface` offset like heatmap? As a continuous thing I would expect it to use `axes(zs, 1)` and `axes(zs, 2)`, not some larger range like heatmap.

Reply:
> Maybe it would be useful to compare to some other plotting libraries? To me it should just be like my after example.

MatPlotLibs `surf` seems to be discrete so it's not comparable. 

## Plots with GR backend
![gr](https://user-images.githubusercontent.com/10947937/230106294-c28d8bec-825f-4483-85b2-24678840fb98.png)

## Plots with Plotly backend
![Screenshot from 2023-04-05 16-06-42](https://user-images.githubusercontent.com/10947937/230106368-00c2c95d-f04f-4c0d-b3a0-671ef9ae41c4.png)

## Plots with PythonPlot backend
![Screenshot from 2023-04-05 16-07-35](https://user-images.githubusercontent.com/10947937/230106494-ce8f68b1-5909-44ff-b861-2111bd973c89.png)

## Plots with Gaston backend
![Screenshot from 2023-04-05 16-16-58](https://user-images.githubusercontent.com/10947937/230109073-5fce944f-f974-4077-afb3-ebfec0a21616.png)

All of these are
```julia
Plots.surface([1 1 1; 2 2 2; 3 3 3], colormap = cgrad([:black, :white, :black], 3))
```
---

## This PR

![Screenshot from 2023-04-05 16-27-42](https://user-images.githubusercontent.com/10947937/230112017-df81cac9-b7a5-4734-9248-80fb8d01e867.png)

Out of those, none do our `0.0:1.5:3.0` x and y range, half do flat face coloring, half do smooth coloring and all match min/max z to the first/last color in the colormap like in this pr.